Add option for `Path3D` to snap to colliders

GustJc opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/11650

Add Path3D option to snap points to colliders to improve creating and editing paths.

Newly created points or moving points will snap the position to the collider intersection instead of the camera plane when the option is enabled.

https://github.com/user-attachments/assets/6b87cccf-23c6-4c07-9e77-353862bbf0fb

You can toggle this behavior on the path3d options menu
<img width="454" height="126" alt="image" src="https://github.com/user-attachments/assets/2c300261-96e2-421b-bcc3-04f13c1142f2" />

----
**Update:**
Fixed to make it also work when physics is in a separate thread.

ryevdokimov reviewed (commented):
This will crash if [physics runs on a separate thread](https://docs.godotengine.org/en/latest/classes/class_projectsettings.html#class-projectsettings-property-physics-3d-run-on-separate-thread), check `NOTIFICATION_PHYSICS_PROCESS` in `Node3DEditorViewport` to see how this was handled for similar functionality.

GustJc commented:
> I would raise the dots a bit using Normal collision

That's a good idea, but it would **have to be a new configuration options to specify by how much**, if any to offset the position.

The options menu (PopUpMenu class) doesn't accept an input field to put the offset value there and looking around it seems the go-to way is to create a confirmation dialogue like the 2D snap settings here.

![_example_snap](https://github.com/user-attachments/assets/e74b3194-b7a6-43b1-b86c-9722fdafdc3a)

I think its best this PR just focus on adding the **basic functionality first**.
And add extra options on another PR, we can then discuss what other options would be helpful too.
The waterways addon for example, gives several snap options to chose from

![image](https://github.com/user-attachments/assets/d02f7e47-d174-44f0-8604-e04bbd1834eb)

geowarin reviewed (approved):
It works great! Nice QOL feature.

![image](https://github.com/user-attachments/assets/49cb631e-5e2b-430c-a201-822495876048)

Thanks @GustJc for this PR and for its accompanying Youtube video (Create a NEW feature to Godot (C++) | How to contribute code), showing people the process of contributing to the engine.

Handling threaded physics adds a little bit of complexity and duplication but it seems reasonable.

In the future, I think it would be nice if "snap to colliders" also worked when dragging an existing point. Maybe in another PR?

GustJc commented:
> Handling threaded physics adds a little bit of complexity and duplication but it seems reasonable.
> 
> In the future, I think it would be nice if "snap to colliders" also worked when dragging an existing point. Maybe in another PR?

Oh, you are right, it totally skipped my mind about moving the handles.
**I've made snap to colliders work while moving the handles now.**

https://github.com/user-attachments/assets/aaf3bf5f-72f6-474b-9bcc-36ad0af26372

But it turns out having support for '3D physics run on separate thread" is proving to be more trouble than it's worth.
There is not easy way to do it because of how Gizmos work with the 'set_handle()' and '_commit_handle()'.
I've tried a few things but there is no easy way to fix it.

And looking at the addons that re-implement their own path3d like 'waterways' and 'proton scatter', it seems they also don't work with "Physics Run on Separate Thread".

Maybe it can be **considered a bug that the editor itself has physics ran on a separate thread**? The setting itself only affects the editor when you change the option and restart it.

Right now, I've made it so there is an error message to at least notify the user.
And it works as usual, if you don't have the snap_to_collider enabled.

https://github.com/user-attachments/assets/6667cd9b-8f36-4866-bedf-8a7abb1f4668

I've left the code that enables adding points with physics on a separate thread. 
But I can also remove it to only work with the main thread like before. It's a cleaner solution.

----
tl;dr: Physics "Run on a Separate Thread" is problematic within the editor. Addons tha re-implement path3d with snap colliders also don't support it.
There seems to be no way to make it work within gizmo's set_handle() right now.
So I left an error message if you try to move the handles and the editor has 'physics on separate thread' enabled.

GustJc commented:
I've rebased and cleaned up the code.
I decided to **remove** the half-support for physics running on a separate thread.
It added a lot of confusing and duplicated code.
And there is no way for Godot to execute a raycast inside a gizmo's set_handle when using physics on threads.

So a cleaner solution is to just to not allow the snap feature if the editor is running physics on threads for now. 
If there is a solution for this in the future, it can be added later.

So for now, I think this PR is done. Awaiting review.

Here's what it looks like if you have physics running on threads:

![snap_path_disabled](https://github.com/user-attachments/assets/9b4b809b-0c8f-4f4e-b397-672d22b9ae61)

----
I have some thoughts about the editor itself being able to run physics on a separate thread, but Godot allows it, and it completely breaks gizmos that needs to use raycast for handles as far as I can tell.

nargleflex commented:
Would really love to see this in soon, I've got terrain that I'm using paths to generate collision areas for and it would be really nice to not have to manually align each point in the curve

AThousandShips commented on editor/scene/3d/path_3d_editor_plugin.h:
```diff
@@ -162,7 +163,8 @@ class Path3DEditorPlugin : public EditorPlugin {
 		HANDLE_OPTION_ANGLE,
-		HANDLE_OPTION_LENGTH
+		HANDLE_OPTION_LENGTH,
+		HANDLE_OPTION_SNAP_COLLIDER
```
```suggestion
		HANDLE_OPTION_SNAP_COLLIDER,
```

ryevdokimov reviewed (changes requested):
If `Snap to Colliders` is checked, and you try to add a point to empty space, nothing happens which is confusing. It should fall back to plane intersection, which is how similar functionality handles it (collision repositioning or instantiating).

GustJc committed:
Add option for Path3D to snap to colliders

ryevdokimov reviewed (approved):
Works, and code looks good to me.

Calinou commented on editor/scene/3d/path_3d_editor_plugin.cpp:
```diff
@@ -884,6 +924,58 @@ void Path3DEditorPlugin::_notification(int p_what) {
+			if (_edit.waiting_handle_physics) {
+				_edit.in_physics_frame = true;
+
+				// Find gizmo reference
```
```suggestion
				// Find gizmo reference.
```

Calinou commented on editor/scene/3d/path_3d_editor_plugin.cpp:
```diff
@@ -114,6 +114,29 @@ void Path3DGizmo::set_handle(int p_id, bool p_secondary, Camera3D *p_camera, con
+			Path3DEditorPlugin::singleton->_edit.gizmo_camera = p_camera;
+			Path3DEditorPlugin::singleton->_edit.mouse_pos = p_point;
+			return;
+			// Only advances this code if inside physics frame and waiting for physics
```
```suggestion
			// Only continue if inside physics frame and waiting for physics.
```

Calinou reviewed (approved):
Tested locally with CSG nodes that have collision enabled, it works as expected. Code looks good to me.
